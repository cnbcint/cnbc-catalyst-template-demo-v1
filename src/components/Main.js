import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { getMargin, breakpoints } from "../helpers/Layout";
import { ModalContext } from "../components/Modal";

export const LayoutContext = React.createContext({
  maxWidth: "500px",
  colors: null,
});

const Container = styled.div`
  display: block;
  width: 100%;
  background-color: ${(props) => props.colors.background};
  > * {
    font-size: 1.125rem;
    line-height: 140%;
    color: ${(props) => props.colors.copy};
  }
  hr {
    // border-color: ${(props) => props.primaryColor};
    border-color: ${(props) => props.colors.copy};
    border-style: solid;
    display: inline-block;
    width: 100%;
    // margin-top: ${getMargin(1)};
    margin-bottom: ${getMargin(1)};
    @media (min-width: ${breakpoints.mobile}) {
      // margin-top: ${getMargin(2)};
      margin-bottom: ${getMargin(2)};
    }
  }
  img {
    width: 100%;
    display: block;
    margin-top: ${getMargin(2)};
    margin-bottom: ${getMargin(2)};
  }
  p {
    margin: 1rem auto !important;
  }
  ul {
    list-style: disc !important;
    margin-top: ${getMargin(1)} !important;
    margin-bottom: ${getMargin(1)} !important;
    li {
      list-style: disc !important;
      margin: ${getMargin(1)} !important;
      margin-left: ${getMargin(2)} !important;
    }
  }
`;
export default function Main(props) {
  return (
    <LayoutContext.Provider
      value={{ maxWidth: props.maxWidth, colors: props.colors }}
    >
      <ModalContext.Provider value={{ modals: {} }}>
        <Container
          colors={props.colors}
          primaryColor={props.colors.primary}
          className={props.className}
          style={props.style}
        >
          {props.children}
        </Container>
      </ModalContext.Provider>
    </LayoutContext.Provider>
  );
}
Main.defaultProps = {
  maxWidth: "900px",
  colors: { primary: "black", secondary: "gray", tertiary: "red" },
};
Main.propTypes = {
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,
  maxWidth: PropTypes.string,
  colors: PropTypes.shape({
    primary: PropTypes.string,
    secondary: PropTypes.string,
    tertiary: PropTypes.string,
    background: PropTypes.string,
    copy: PropTypes.string,
  }),
  style: PropTypes.object,
};
