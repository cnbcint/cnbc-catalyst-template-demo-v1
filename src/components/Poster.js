import React, { useContext } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import useAnimation from "../hooks/useAnimation";
import {
  breakpoints,
  getAnimation,
  getMargin,
  getJustifyContent,
} from "../helpers/Layout";
import { LayoutContext } from "./Main";

const Container = styled.div`
  font-size: 1.5rem;
  @media (min-width: ${breakpoints.mobile}) {
    font-size: 2rem;
  }
  background-color: none;
  line-height: 150%;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  margin: 0;
  box-sizing: border-box;
  text-align: ${(props) => props.aligment};
`;

const Picture = styled.picture`
  display: block;
  img {
    margin: 0;
  }
`;
const Content = styled.div`
  position: absolute;
  width: 100%;
  @media (min-width: ${breakpoints.mobile}) {
    width: 80%;
  }
  padding: ${getMargin(2)};
  box-sizing: border-box;
  color: white;
  display: flex;
  flex-direction: column;
  align-items: ${(props) => getJustifyContent(props.aligment)};
  > hr {
    width: 50%;
    margin: ${getMargin(2)} 0;
    border-color: ${(props) => props.dividerColor};
    background-color: ${(props) => props.dividerColor};
  }
  text-align: ${(props) => props.aligment};
`;

const Overlay = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.75);
`;
/**
 Component that allows to have a quote or a sentence with background image.
*/
export default function Poster(props) {
  const layout = useContext(LayoutContext);
  const { ref: refCopy } = useAnimation({
    animationName: getAnimation(props.aligment),
    duration: 500,
    delay: 0,
    loop: props.loopAnimations,
    autoStart: true,
  });
  const { ref: refDevider } = useAnimation({
    animationName: "fadeIn",
    duration: 500,
    delay: 125,
    loop: props.loopAnimations,
    autoStart: true,
  });
  return (
    <Container
      className={props.className}
      aligment={props.aligment}
      style={props.style}
    >
      {props.image && (
        <Picture>
          <source
            media={`(min-width:${breakpoints.tablet})`}
            srcSet={props.image.desktop}
          />
          <source
            media={`(min-width:${breakpoints.mobile})`}
            srcSet={props.image.tablet}
          />
          <img src={props.image.mobile} alt="poster" loading="lazy" />
        </Picture>
      )}
      <Overlay />
      <Content
        className={props.className ? props.className + "__content" : ""}
        aligment={props.aligment}
        dividerColor={layout.colors[props.dividerColor]}
      >
        <div ref={refCopy}>{props.children}</div>
        {props.addDivider && <hr ref={refDevider} />}
      </Content>
    </Container>
  );
}

Poster.defaultProps = {
  dividerColor: "primary",
  loopAnimations: true,
  aligment: "center",
  className: "",
  style: {},
  addDivider: true,
};

Poster.propTypes = {
  /**
   * Allow to hide or show divider line after the text
   */
  addDivider: PropTypes.bool,
  /**
   * Allow to chosse a devider colors from Theme using the 3 types of colors
   */
  dividerColor: PropTypes.oneOf(["primary", "secondary", "tertiary"]),
  className: PropTypes.string,
  /**
   * React CSS fromat for overriding styles
   */
  style: PropTypes.object,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,
  /**
   * Allow animations to play gain once the user scrolls this element back into the screen
   */
  loopAnimations: PropTypes.bool,
  /**
   * Image into responsive format with 3 breakpoints
   */
  image: PropTypes.shape({
    mobile: PropTypes.string,
    tablet: PropTypes.string,
    desktop: PropTypes.string,
  }),
  /**
   * This also affect animation
   */
  aligment: PropTypes.oneOf(["left", "center", "right"]),
};
