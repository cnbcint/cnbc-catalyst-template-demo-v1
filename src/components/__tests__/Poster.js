import React from "react";
import Poster from "../Poster";
import Main from "../Main";
import { render } from "@testing-library/react";
import "jest-styled-components";

const Theme = {
  maxWidth: "900px",
  colors: {
    primary: "orange",
    secondary: "green",
    tertiary: "orange",
    background: "white",
    copy: "black",
    links: "blue",
  },
};

const Component = () => (
  <Main {...Theme}>
    <Poster
      image={{
        desktop:
          "http://localhost:8080/images/responsive/output/image05-desktop.jpg",
        tablet:
          "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
        mobile:
          "http://localhost:8080/images/responsive/output/image05-mobile.jpg",
      }}
      aligment="right"
      className="test"
    >
      Testing Content
    </Poster>
  </Main>
);

beforeAll(() => {
  window.IntersectionObserver = class {
    observe() {
      return null;
    }
    disconnect() {}
  };
});
describe("Poster", () => {
  it("Renders correctly", () => {
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Renders an img with default mobile source", () => {
    const { container } = render(<Component />);
    expect(container.querySelector("img").src).toMatch(
      "http://localhost:8080/images/responsive/output/image05-mobile.jpg"
    );
  });

  it("Add sources for desktop and tablet", () => {
    const { container } = render(<Component />);
    expect(container.querySelectorAll("source")[0].srcset).toMatch(
      "http://localhost:8080/images/responsive/output/image05-desktop.jpg"
    );
    expect(container.querySelectorAll("source")[1].srcset).toMatch(
      "http://localhost:8080/images/responsive/output/image05-tablet.jpg"
    );
  });

  it("Sources has the correct media queries for tablet and desktop", () => {
    const { container } = render(<Component />);
    expect(container.querySelectorAll("source")[0].media).toMatch(
      "(min-width:1280px)"
    );
    expect(container.querySelectorAll("source")[1].media).toMatch(
      "(min-width:768px)"
    );
  });

  it("Renders the correct content", () => {
    const { container } = render(<Component />);
    expect(
      container.getElementsByClassName("test__content")[0].textContent
    ).toMatch("Testing Content");
  });

  it("Renders correctly with no dividers", () => {
    const { container } = render(<Component addDivider={false} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Passes the class name", () => {
    const { container } = render(<Component />);
    expect(container.firstChild.firstChild.classList.contains("test")).toBe(
      true
    );
  });

  it("Set The right aligment", () => {
    const { container } = render(<Component />);
    expect(
      container.getElementsByClassName("test__content")[0]
    ).toHaveStyleRule("text-align", "right");
  });
});

afterAll(() => {
  window.IntersectionObserver = null;
});
