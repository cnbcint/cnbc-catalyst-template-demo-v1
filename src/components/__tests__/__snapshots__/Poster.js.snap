// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poster Renders correctly 1`] = `
.c0 {
  display: block;
  width: 100%;
  background-color: white;
}

.c0 > * {
  font-size: 1.125rem;
  line-height: 140%;
  color: black;
}

.c0 hr {
  border-color: black;
  border-style: solid;
  display: inline-block;
  width: 100%;
  margin-bottom: 1rem;
}

.c0 img {
  width: 100%;
  display: block;
  margin-top: 2rem;
  margin-bottom: 2rem;
}

.c0 p {
  margin-top: 1rem !important;
  margin-bottom: 1rem !important;
}

.c0 ul {
  list-style: disc !important;
  margin-top: 1rem !important;
  margin-bottom: 1rem !important;
}

.c0 ul li {
  list-style: disc !important;
  margin: 1rem !important;
  margin-left: 2rem !important;
}

.c1 {
  font-size: 1.5rem;
  background-color: none;
  line-height: 150%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin: 0;
  box-sizing: border-box;
  text-align: right;
}

.c2 {
  display: block;
}

.c2 img {
  margin: 0;
}

.c4 {
  position: absolute;
  width: 100%;
  padding: 2rem;
  box-sizing: border-box;
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  text-align: right;
}

.c4 > hr {
  width: 50%;
  margin: 2rem 0;
  border-color: orange;
  background-color: orange;
}

.c3 {
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: rgba(0,0,0,0.75);
}

@media (min-width:768px) {
  .c0 hr {
    margin-bottom: 2rem;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 2rem;
  }
}

@media (min-width:768px) {
  .c4 {
    width: 80%;
  }
}

<div
  class="c0"
>
  <div
    class="c1 test"
  >
    <picture
      class="c2"
    >
      <source
        media="(min-width:1280px)"
        srcset="http://localhost:8080/images/responsive/output/image05-desktop.jpg"
      />
      <source
        media="(min-width:768px)"
        srcset="http://localhost:8080/images/responsive/output/image05-tablet.jpg"
      />
      <img
        alt="poster"
        loading="lazy"
        src="http://localhost:8080/images/responsive/output/image05-mobile.jpg"
      />
    </picture>
    <div
      class="c3"
    />
    <div
      class="c4 test__content"
    >
      <div
        class="pre-animated"
      >
        Testing Content
      </div>
      <hr
        class="pre-animated"
      />
    </div>
  </div>
</div>
`;

exports[`Poster Renders correctly with no dividers 1`] = `
.c0 {
  display: block;
  width: 100%;
  background-color: white;
}

.c0 > * {
  font-size: 1.125rem;
  line-height: 140%;
  color: black;
}

.c0 hr {
  border-color: black;
  border-style: solid;
  display: inline-block;
  width: 100%;
  margin-bottom: 1rem;
}

.c0 img {
  width: 100%;
  display: block;
  margin-top: 2rem;
  margin-bottom: 2rem;
}

.c0 p {
  margin-top: 1rem !important;
  margin-bottom: 1rem !important;
}

.c0 ul {
  list-style: disc !important;
  margin-top: 1rem !important;
  margin-bottom: 1rem !important;
}

.c0 ul li {
  list-style: disc !important;
  margin: 1rem !important;
  margin-left: 2rem !important;
}

.c1 {
  font-size: 1.5rem;
  background-color: none;
  line-height: 150%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin: 0;
  box-sizing: border-box;
  text-align: right;
}

.c2 {
  display: block;
}

.c2 img {
  margin: 0;
}

.c4 {
  position: absolute;
  width: 100%;
  padding: 2rem;
  box-sizing: border-box;
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  text-align: right;
}

.c4 > hr {
  width: 50%;
  margin: 2rem 0;
  border-color: orange;
  background-color: orange;
}

.c3 {
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: rgba(0,0,0,0.75);
}

@media (min-width:768px) {
  .c0 hr {
    margin-bottom: 2rem;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 2rem;
  }
}

@media (min-width:768px) {
  .c4 {
    width: 80%;
  }
}

<div
  class="c0"
>
  <div
    class="c1 test"
  >
    <picture
      class="c2"
    >
      <source
        media="(min-width:1280px)"
        srcset="http://localhost:8080/images/responsive/output/image05-desktop.jpg"
      />
      <source
        media="(min-width:768px)"
        srcset="http://localhost:8080/images/responsive/output/image05-tablet.jpg"
      />
      <img
        alt="poster"
        loading="lazy"
        src="http://localhost:8080/images/responsive/output/image05-mobile.jpg"
      />
    </picture>
    <div
      class="c3"
    />
    <div
      class="c4 test__content"
    >
      <div
        class="pre-animated"
      >
        Testing Content
      </div>
      <hr
        class="pre-animated"
      />
    </div>
  </div>
</div>
`;
